fix: close RedisStore client in case the with_client classmethod is used

euri10 opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->
## Description

WIP, it doesn't pass the test I added from the mvce in the issue (this one could certainly be simplified) so my impleentation has obviously an issue.
I made the `RedisStore` an async context manager, I can see the `_on_shutdon` method being hit, but the `_redis` client is still not closed or it doesnt exists.

from your comment here:
https://discord.com/channels/919193495116337154/1152626430295933060/1207016578823102565


<!--
Please add in issue numbers this pull request will close, if applicable
Examples: Fixes #4321 or Closes #1234


Ensure you are using a supported keyword to properly link an issue:
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
## Closes


Fixes #3083

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -64,6 +68,20 @@ def __init__(self, redis: Redis, namespace: str | None | EmptyType = Empty) -> N
+    async def _on_shutdown(self) -> None:
+        await self._redis.aclose()
+
+    async def __aenter__(self) -> RedisStore:
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        await self._on_shutdown()
```
These should be added to the base store and simply do nothing

provinzkraut commented on litestar/app.py:
```diff
@@ -465,6 +465,9 @@ def __init__(
         self.stores: StoreRegistry = (
             config.stores if isinstance(config.stores, StoreRegistry) else StoreRegistry(config.stores)
         )
+        for _, store in self.stores._stores.items():
+            if not store._should_close_store_manually:
+                self._lifespan_managers.append(store)
```
If you have implemented the context manager methods on the store base class.
```suggestion
        for store in self.stores._stores.values():
           self._lifespan_managers.append(store)
```

I'd also move the store registration and this handling logic somewhere within the init before the `self._lifespan_managers` are initialised, so you can simply append it to `config.lifespan`

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -30,6 +33,7 @@ def __init__(self, redis: Redis, namespace: str | None | EmptyType = Empty) -> N
                 ``None``. This will make :meth:`.delete_all` unavailable.
         """
         self._redis = redis
+        self._should_close_store_manually = True
```
Maybe have this as a parameter in `__init__`, so you don't need to patch this from the outside?

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -68,19 +73,19 @@ def __init__(self, redis: Redis, namespace: str | None | EmptyType = Empty) -> N
 
-    async def __aenter__(self) -> RedisStore:
-        return self
+    async def __aenter__(self) -> None:
+        ...
```
```suggestion
```

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -18,23 +18,28 @@
 
-    def __init__(self, redis: Redis, namespace: str | None | EmptyType = Empty) -> None:
+    def __init__(
+        self, redis: Redis, namespace: str | None | EmptyType = Empty, automatic_store_lifetime: bool = False
+    ) -> None:
         """Initialize :class:`RedisStore`
 
         Args:
             redis: An :class:`redis.asyncio.Redis` instance
             namespace: A key prefix to simulate a namespace in redis. If not given,
                 defaults to ``LITESTAR``. Namespacing can be explicitly disabled by passing
                 ``None``. This will make :meth:`.delete_all` unavailable.
+            automatic_store_lifetime: If ``True``, the store lifetime will be handled automatically,
+            works only if you're using the :meth:`.with_client` method.
+            By default it is False and in that case you will have to handle the store lifetime manually.
```
```suggestion
        self, redis: Redis, namespace: str | None | EmptyType = Empty, handle_client_shutdown: bool = False
    ) -> None:
        """Initialize :class:`RedisStore`

        Args:
            redis: An :class:`redis.asyncio.Redis` instance
            namespace: A key prefix to simulate a namespace in redis. If not given,
                defaults to ``LITESTAR``. Namespacing can be explicitly disabled by passing
                ``None``. This will make :meth:`.delete_all` unavailable.
            handle_client_shutdown: If ``True``, handle the shutdown of the `redis` instance automatically during the store's lifespan. Should be set to `True` unless the shutdown is handled externally
```

I think this way is a bit more clear

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -92,6 +97,7 @@ def with_client(
         username: str | None = None,
         password: str | None = None,
         namespace: str | None | EmptyType = Empty,
+        automatic_store_lifetime: bool = True,
```
```suggestion
```

I don't think we need this option here; If we create the client, we should always close it properly

euri10 commented:
you're reviewing / holding mye hand faster than I can type :)
but there's progress, the test I wrote from the mcve passes,
I'll check later on why it broke 46 existing ones :+1:

euri10 commented on litestar/stores/base.py:
```diff
@@ -76,6 +78,19 @@ async def expires_in(self, key: str) -> int | None:
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        ...
```
just had to ignore the B027 to please the linter

provinzkraut commented:
> you're reviewing / holding mye hand faster than I can type :) 

Sorry :grimacing:

euri10 commented:
> > you're reviewing / holding mye hand faster than I can type :)
> 
> Sorry 😬

no worries that was fun, I almost wondered if you were not a LLM at sone point :rofl:

provinzkraut commented on litestar/app.py:
```diff
@@ -388,7 +388,13 @@ def __init__(
+        )
         self._lifespan_managers = config.lifespan
+        for store in self.stores._stores.values():
+            if store.handle_client_shutdown:
+                self._lifespan_managers.append(store)
```
I think my earlier suggestion was bad. 
What I meant was:

- Always add the stores as lifespan managers here
- Handle the `handle_client_shutdown` inside `RedisStore.__aexit__`

That way each store can decide how it wants to handle shutdown if at all

euri10 replied:
gotcha, I'll take that back later on, you obviously know in and out the codebase, that's very impressive, the LLM-joke was meant as a compliment by the way !

euri10 replied:
> I think my earlier suggestion was bad. What I meant was:
> 
> * Always add the stores as lifespan managers here
> * Handle the `handle_client_shutdown` inside `RedisStore.__aexit__`
> 
> That way each store can decide how it wants to handle shutdown if at all

done, it's better since it clears those 46 breaks in tests by keeping existing stores the same.

also, the current test is like....complicated, but in the meantime the fact it popped up for me is exactly how we figured the closing wasn't performed, so I'd be tempted to keep it as is, but happ

time for some linter stuff now

provinzkraut commented:
> no worries that was fun, I almost wondered if you were not a LLM at sone point 🤣

I do wonder that too at times. Aren't we all just fancy LLMs at some level? :eyes:

euri10 commented:
> > no worries that was fun, I almost wondered if you were not a LLM at sone point 🤣
> 
> I do wonder that too at times. Aren't we all just fancy LLMs at some level? 👀

that kind of would be Nick Bolstrom's theory in "Are you living in a computer simulation?"

euri10 commented:
I might have introduced some flakiness :)
Will take a look tomorrow

euri10 commented:
> I might have introduced some flakiness :) Will take a look tomorrow

ok in fact it's just the test I added doesnt integrate well in the suite, need to find a way to see how to keep it without issues

euri10 commented:
ok I removed my initial test from the mcve that passes with the change made to the lifespan, reason is twofold.

one, as you can notice in the suite, all instances where with_client is used are patched, so if there is a test to be made, one has to follow the same rule or we may end up having unclosed ressources: I therefore think this test is problematic: https://github.com/litestar-org/litestar/blob/323e23e51508cf142074447098b48ddcc2a67ad0/tests/unit/test_stores.py#L242-L245
because it doesnt patch the with_client and will leave things unclosed, the other line is fine

Also, I modified the redis_client fixture: https://github.com/litestar-org/litestar/pull/3111/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128L307-R315

the previous one is problematic because it uses a sync redis instance to flushall and never closes it, one can just use the async one from the fixture and flushall in the finally for the same effect without the unintended consequences,

wdyt ?

provinzkraut commented:
> the previous one is problematic because it uses a sync redis instance to flushall and never closes it, one can just use the async redis instance from the fixture and flushall in the finally for the same effect without the unintended consequences,
> 
> wdyt ?

I think the proper fix would be to close the sync `Redis` instance. The reason we flush at the beginning and not the end is that if we're flushing at the end, and something goes wrong, we'd be affecting other tests. If we flush at the beginning, every test *starts* with an empty DB. 

Also, I'm surprised that calling `.flush()` after `.close()` does not cause any issues.

euri10 commented:
> > the previous one is problematic because it uses a sync redis instance to flushall and never closes it, one can just use the async redis instance from the fixture and flushall in the finally for the same effect without the unintended consequences,
> > wdyt ?
> 
> I think the proper fix would be to close the sync `Redis` instance. The reason we flush at the beginning and not the end is that if we're flushing at the end, and something goes wrong, we'd be affecting other tests. If we flush at the beginning, every test _starts_ with an empty DB.
> 
> Also, I'm surprised that calling `.flush()` after `.close()` does not cause any issues.

so @provinzkraut I sent a PR fixing the fixture think properly I think here: https://github.com/litestar-org/litestar/pull/3111

questions now on this one:
1. do we deal here with the only test using with_client that doesnt uses mocks.
2. should I add a test checking if the client is closed when using that classmethod, provided there is no is_closed method on the redis client, so I wouldn't know how to test it

provinzkraut commented on tests/conftest.py:
```diff
@@ -304,12 +304,7 @@ def get_logger() -> GetLogger:
+async def redis_client(docker_ip: str, redis_service: None) -> AsyncRedis:
+    async with AsyncRedis(host=docker_ip, port=6397) as client:
         await client.flushall()
+        return client
```
I don't understand why this works. The context manager closes the client immediately before returning so won't the client be closed for all the tests?

euri10 replied:
now you mention it this is weird indeed,
https://github.com/litestar-org/litestar/pull/3117 is the change for the redis fixture and all tests pass fine, yielding the client and nothing works,

euri10 replied:
given https://docs.python.org/3/reference/compound_stmts.html#async-with

we have then
```
    async with AsyncRedis(host=docker_ip, port=6397) as client:
        await client.flushall()
        return client
```
is in fact (I may have syntax errors here it's just to understand the flow)

```
hit_except = False

try:
    client = AsyncRedis.__aenter__()
    await client.flushall()
    return client
except:
    hit_except = True
    if not await aexit(manager, *sys.exc_info()):
        raise
finally:
    if not hit_except:
        await AsyncRedis.__aexit()
```

provinzkraut commented on litestar/stores/redis.py:
```diff
@@ -84,6 +103,7 @@ def with_client(
             username: Redis username to use
             password: Redis password to use
             namespace: Virtual key namespace to use
+            automatic_store_lifetime: If ``True``, the store lifetime will be handled automatically, it is the default with this method.
```
```suggestion
```

euri10 commented on litestar/stores/redis.py:
```diff
@@ -93,14 +112,22 @@ def with_client(
         The current instances namespace will serve as a prefix for the namespace, so it
         can be considered the parent namespace.
         """
-        return type(self)(redis=self._redis, namespace=f"{self.namespace}_{namespace}" if self.namespace else namespace)
```
I didn't add a test on this as I wouldn't know how to do this, the reason it is necessary is that if you are using a namedspace store and with_client then in tests this attribute will get overriden

euri10 commented:
I think it should be ready, I'm not pleased with the fixtures as-is to be fair, so I reverted my changes on this part but that shouldn't prevent the PR to be looked at ;)
